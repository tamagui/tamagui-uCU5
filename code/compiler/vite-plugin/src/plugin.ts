import type { TamaguiOptions } from '@tamagui/static'
import * as StaticIn from '@tamagui/static'
import type { Plugin } from 'vite'
import { transformWithEsbuild } from 'vite'

const Static = (StaticIn['default'] || StaticIn) as typeof StaticIn

export function tamaguiPlugin(tamaguiOptionsIn: TamaguiOptions = {}): Plugin {
  const options = {
    ...tamaguiOptionsIn,
    ...Static.loadTamaguiBuildConfigSync(tamaguiOptionsIn),
  }

  const { platform = 'web' } = options

  const watcher = options.disableWatchTamaguiConfig
    ? null
    : Static.watchTamaguiConfig({
        platform,
        components: ['tamagui'],
        config: './src/tamagui.config.ts',
        ...options,
      }).catch((err) => {
        console.error(` [Tamagui] Error watching config: ${err}`)
      })

  const components = [
    ...new Set([...(options.components || []), 'tamagui', '@tamagui/core']),
  ]
  const noExternalSSR = new RegExp(
    `${components.join('|')}|react-native|expo-linear-gradient`,
    'ig'
  )

  const extensions = [
    `.${platform}.mjs`,
    `.${platform}.js`,
    `.${platform}.jsx`,
    `.${platform}.ts`,
    `.${platform}.tsx`,
    '.mjs',
    '.js',
    '.mts',
    '.ts',
    '.jsx',
    '.tsx',
    '.json',
  ]

  const plugin: Plugin = {
    name: 'tamagui-base',
    enforce: 'pre',

    async buildEnd() {
      await watcher?.then((res) => {
        res?.dispose()
      })
    },

    // // fix expo-linear-gradient
    async transform(code, id) {
      if (!id.includes('expo-linear-gradient')) {
        return
      }
      // Use the exposed transform from vite, instead of directly
      // transforming with esbuild
      return transformWithEsbuild(code, id, {
        loader: 'jsx',
        jsx: 'automatic', // ðŸ‘ˆ this is important
      })
    },

    async config(userConfig, env) {
      return {
        define: {
          // reanimated support
          _frameTimestamp: undefined,
          _WORKLET: false,
          __DEV__: `${env.mode === 'development' ? true : false}`,
          'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV || env.mode),
          'process.env.ENABLE_RSC': JSON.stringify(process.env.ENABLE_RSC || ''),
          'process.env.ENABLE_STEPS': JSON.stringify(process.env.ENABLE_STEPS || ''),
          'process.env.IS_STATIC': JSON.stringify(false),
          ...(env.mode === 'production' && {
            'process.env.TAMAGUI_OPTIMIZE_THEMES': JSON.stringify(true),
          }),
        },
        optimizeDeps: {
          jsx: 'transform',
          include: platform === 'web' ? ['expo-linear-gradient'] : [],
          // disabled: false,
          esbuildOptions: {
            resolveExtensions: extensions,
            loader: {
              '.js': 'jsx',
            },
          },
        },
        ssr: {
          noExternal: noExternalSSR,
        },
        resolve: {
          extensions: extensions,
          alias: {
            ...(platform !== 'native' && {
              'react-native/Libraries/Renderer/shims/ReactFabric': '@tamagui/proxy-worm',
              'react-native/Libraries/Utilities/codegenNativeComponent':
                '@tamagui/proxy-worm',
              'react-native-svg': '@tamagui/react-native-svg',
              'react-native': 'react-native-web',
              ...(options.useReactNativeWebLite && {
                'react-native': 'react-native-web-lite',
                'react-native-web': 'react-native-web-lite',
              }),
              ...(options.useReactNativeWebLite === 'without-animated' && {
                'react-native': 'react-native-web-lite/without-animated',
                'react-native-web': 'react-native-web-lite/without-animated',
              }),
            }),
          },
        },
      }
    },
  }

  return plugin
}
