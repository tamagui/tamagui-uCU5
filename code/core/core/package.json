{
  "name": "@tamagui/core",
  "version": "1.102.3",
  "source": "src/index.tsx",
  "main": "dist/cjs",
  "module": "dist/esm",
  "tamagui": {
    "bundle.native": "./src/index.tsx",
    "bundle.native.test": "./src/index.tsx"
  },
  "types": "./types/index.d.ts",
  "license": "MIT",
  "author": {
    "name": "Nate Wienert"
  },
  "sideEffects": [
    "*.css"
  ],
  "scripts": {
    "build": "tamagui-build",
    "watch": "tamagui-build --watch",
    "lint": "npx biome check src",
    "lint:fix": "npx biome check --apply-unsafe src",
    "clean": "tamagui-build clean",
    "clean:build": "tamagui-build clean:build"
  },
  "files": [
    "types",
    "dist",
    "src",
    "reset.css",
    "native.js",
    "native.d.ts",
    "native-test.js",
    "native-test.d.ts"
  ],
  "dependencies": {
    "@tamagui/react-native-use-pressable": "1.102.3",
    "@tamagui/react-native-use-responder-events": "1.102.3",
    "@tamagui/use-event": "1.102.3",
    "@tamagui/web": "1.102.3"
  },
  "devDependencies": {
    "@tamagui/build": "1.102.3",
    "@testing-library/react": "^14.0.0",
    "csstype": "^3.0.10",
    "typescript": "^5.5.2",
    "vitest": "^0.34.3"
  },
  "exports": {
    "./package.json": "./package.json",
    "./reset.css": "./reset.css",
    ".": {
      "react-native": "./dist/cjs/index.native.js",
      "types": "./types/index.d.ts",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/cjs/index.js"
    },
    "./native": {
      "types": "./types/index.d.ts",
      "import": "./dist/native.js",
      "require": "./dist/native.js"
    },
    "./native-test": {
      "types": "./types/index.d.ts",
      "import": "./dist/test.native.js",
      "require": "./dist/test.native.js"
    },
    "./inject-styles": {
      "react-native": "./dist/cjs/inject-styles.native.js",
      "types": "./types/inject-styles.d.ts",
      "import": "./dist/esm/inject-styles.mjs",
      "require": "./dist/cjs/inject-styles.js"
    },
    "./reactNativeTypes": {
      "react-native": "./dist/cjs/reactNativeTypes.native.js",
      "types": "./types/reactNativeTypes.d.ts",
      "import": "./dist/esm/reactNativeTypes.mjs",
      "require": "./dist/cjs/reactNativeTypes.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/tamagui/tamagui.git",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "a49cc7ea6b93ba384e77a4880ae48ac4a5635c14"
}
