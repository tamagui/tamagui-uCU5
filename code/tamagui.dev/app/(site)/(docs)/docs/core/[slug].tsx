import { ThemeTint } from '@tamagui/logo'
import { useLoader } from 'vxs'
import { getMDXComponent } from 'mdx-bundler/client'
import { useMemo } from 'react'
import { SubTitle, nbspLastWord } from '~/components/SubTitle'
import { DocsQuickNav } from '~/features/docs/DocsQuickNav'
import { HomeH1 } from '~/features/site/home/HomeHeaders'

import { getAllFrontmatter, getMDXBySlug } from '@tamagui/mdx'
import { components } from '~/features/mdx/MDXComponents'
import { HeadInfo } from '~/components/HeadInfo'
import { getOgUrl } from '~/features/site/getOgUrl'

export async function generateStaticParams() {
  const frontmatters = getAllFrontmatter('data/docs/core')
  const paths = frontmatters.map(({ slug }) => ({
    slug: slug.replace(/.*docs\/core\//, ''),
  }))
  return paths
}

export async function loader({ params }) {
  const { frontmatter, code } = await getMDXBySlug('data/docs/core', params.slug)
  return {
    frontmatter,
    code,
  }
}

export default function DocCorePage() {
  const { code, frontmatter } = useLoader(loader)
  const Component = useMemo(() => getMDXComponent(code), [code])

  return (
    <>
      <HeadInfo
        title={`${frontmatter.title}`}
        description={frontmatter.description}
        openGraph={{
          images: [
            {
              url: getOgUrl({
                title: frontmatter.title,
                description: frontmatter.description ?? '',
                category: 'intro',
              }),
            },
          ],
        }}
      />
      <HomeH1>{nbspLastWord(frontmatter.title)}</HomeH1>
      <SubTitle>{nbspLastWord(frontmatter.description || '')}</SubTitle>
      <ThemeTint>
        <Component components={components as any} />
      </ThemeTint>
      <DocsQuickNav />
    </>
  )
}
