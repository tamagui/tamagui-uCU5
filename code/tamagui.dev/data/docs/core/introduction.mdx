---
title: Introduction
description: What is @tamagui/core?
---

<IntroParagraph>
  Core is a React [style and design system library](/docs/core/configuration), a bit like a fully-typed and modern [styled-components](https://styled-components.com) or [Stitches](https://stitches.dev) that can compile to near 0-runtime, and happens to work on React Native the same as on web, despite having no outside dependencies.
</IntroParagraph>

Well, sort of - on iOS and Android it depends on React Native, but on the web it has no outside dependencies besides React - yet it supports the full React Native API surface in 24-28Kb, depending on tree-shaking.

It primarily exports a [`styled`](/docs/core/styled) function and [`View` & `Text`](/docs/core/stack-and-text) components, all of which accept the superset of the React Native API surface that Tamagui supports. 

Tamagui supports a [superset](/docs/intro/styles) of the React Native style API, bringing many powerful abilities from CSS like media queries, pseudo selectors, container queries, and other selectors.

With [`createTamagui`](/docs/core/configuration) you can create what are essentially media queries and typed CSS variables that work for styling, but also with the [`useTheme`](/docs/core/use-theme) and [`useMedia`](/docs/core/use-media) hooks. Both can compile away with the optimizing compiler when used for styling, but they also are written to be extremely performant with explicit dirty tracking and signal-like re-render granularity.

Core supports its full feature-set at runtime and at compile-time (with [`@tamagui/static`](/docs/intro/compiler-install)). This means it's easy to set up and test out, but gives amazing performance when you need it.

Below `@tamagui/core` is `@tamagui/web`, which drops the React Native non-style props and saves an extra 10Kb or so of size. It's a great web-only style library much like a typed version of Stitches.