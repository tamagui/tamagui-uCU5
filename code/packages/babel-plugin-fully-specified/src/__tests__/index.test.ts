import path from 'node:path'
import { describe, expect, test } from 'vitest'
import { type TransformOptions, transform, transformFileSync } from '@babel/core'

import plugin, { type FullySpecifiedOptions } from '../'

/** A helper function to get the default transform options for the plugin to test. */
const getTransformOptions = ({
  pluginOptions,
}: { pluginOptions?: Partial<FullySpecifiedOptions> } = {}) => ({
  plugins: [[plugin, { ensureFileExists: false, ...pluginOptions }]],
})

/** A test helper for calling Babel transform with the plugin to test and some default options. */
function getTransformResult(
  input,
  {
    transformOptions,
    pluginOptions,
  }: {
    transformOptions?: TransformOptions
    pluginOptions?: Partial<FullySpecifiedOptions>
  } = {}
) {
  return transform(input, {
    filename: 'myFile.js',
    configFile: false,
    ...transformOptions,
    ...getTransformOptions({ pluginOptions }),
  })
}

describe('local imports', () => {
  test('named import', () => {
    const example1 = "import { foo } from './foo'"
    expect(getTransformResult(example1)?.code).toBe('import { foo } from "./foo.js";')

    const example2 = "import { foo as bar } from './foo'"
    expect(getTransformResult(example2)?.code).toBe(
      'import { foo as bar } from "./foo.js";'
    )
  })

  test('default import', () => {
    const example1 = "import defaultFoo from './foo'"
    expect(getTransformResult(example1)?.code).toBe('import defaultFoo from "./foo.js";')

    const example2 = "import defaultFoo, { foo } from './foo'"
    expect(getTransformResult(example2)?.code).toBe(
      'import defaultFoo, { foo } from "./foo.js";'
    )
  })

  test('namespace import', () => {
    const example = "import * as fooModule from './foo'"
    expect(getTransformResult(example)?.code).toBe(
      'import * as fooModule from "./foo.js";'
    )
  })

  test('side effects only import', () => {
    const example = "import './foo'"
    expect(getTransformResult(example)?.code).toBe('import "./foo.js";')
  })

  test('dynamic import()', () => {
    const example = "const foo = await import('./foo')"
    expect(getTransformResult(example)?.code).toBe(
      'const foo = await import("./foo.js");'
    )
  })
})

describe('local re-exports', () => {
  test('named', () => {
    const example1 = "export { foo } from './foo'"
    expect(getTransformResult(example1)?.code).toBe('export { foo } from "./foo.js";')

    const example2 = "export { foo as bar } from './foo'"
    expect(getTransformResult(example2)?.code).toBe(
      'export { foo as bar } from "./foo.js";'
    )
  })

  test('default', () => {
    const example1 = "export { default as fooDefault } from './foo'"
    expect(getTransformResult(example1)?.code).toBe(
      'export { default as fooDefault } from "./foo.js";'
    )
  })

  test('namespace', () => {
    const example1 = "export * as fooModule from './foo'"
    expect(getTransformResult(example1)?.code).toBe(
      'export * as fooModule from "./foo.js";'
    )

    const example2 = "export * from './foo'"
    expect(getTransformResult(example2)?.code).toBe('export * from "./foo.js";')
  })
})

describe('transforming actual files', () => {
  test('test', () => {
    const { code } =
      transformFileSync(
        path.join(__dirname, 'fixtures', 'sample-project-1', 'test.mjs'),
        getTransformOptions({
          pluginOptions: { ensureFileExists: true, includePackages: ['@my-org/my-pkg'] },
        })
      ) || {}

    expect(code).toMatchSnapshot()
  })

  test('multiple extensions exists', () => {
    const { code } =
      transformFileSync(
        path.join(__dirname, 'fixtures', 'multiple-extensions-exists', 'test.mjs'),
        getTransformOptions({
          pluginOptions: { ensureFileExists: true },
        })
      ) || {}

    expect(code).toMatchSnapshot()
  })

  test('force extension', () => {
    const { code } =
      transformFileSync(
        path.join(__dirname, 'fixtures', 'force-extension', 'foo.js'),
        getTransformOptions({
          pluginOptions: { ensureFileExists: { forceExtension: '.mjs' } },
        })
      ) || {}

    expect(code).toBe('import { bar } from "./bar.mjs";')
  })
})
