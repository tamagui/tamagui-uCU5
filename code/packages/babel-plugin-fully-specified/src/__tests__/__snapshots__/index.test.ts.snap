// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transforming actual files > multiple extensions exists 1`] = `
"// Expected to be transformed to use the \`.mjs\` extension, which is same as this file.
import { foo, bar } from \\"./modules/module.mjs\\";"
`;

exports[`transforming actual files > test 1`] = `
"import foo, { bar } from \\"./modules/module.mjs\\";
import { foo as foo2 } from \\"./modules/module.mjs\\";
import * as someModule from \\"./modules/module.mjs\\";
import { foo as cjsFoo } from \\"./modules/cjs-module.cjs\\";

// Will not be transformed
import { foo as packageFoo } from '@my-org/my-pkg';

// These will be transformed if \`includePackages\` includes \`@my-org/my-pkg\`.
import { someStuff } from \\"@my-org/my-pkg/lib/index.js\\";
// import { someOtherStuff } from '@my-org/my-pkg/lib/index' // TODO: Fix me
import { exampleFunction } from \\"@my-org/my-pkg/lib/exampleFunction.js\\";"
`;
